src/Interpreters/GraceHashJoin.cpp

```diff
@@ -124,15 +127,14 @@ class GraceHashJoin::FileBucket : boost::noncopyable
         TemporaryFileStream::Stat right;
     };
 
-    explicit FileBucket(size_t bucket_index_,
-                        TemporaryFileStream & left_file_,
-                        TemporaryFileStream & right_file_,
-                        Poco::Logger * log_)
+    explicit FileBucket(size_t bucket_index_, GraceHashJoin & grace_join_)
         : idx{bucket_index_}
-        , left_file{left_file_}
-        , right_file{right_file_}
+        , left_file{
+            grace_join_.tmp_data->createStream(grace_join_.left_sample_block)}
+        , right_file{
+            grace_join_.tmp_data->createStream(grace_join_.prepareRightBlock(grace_join_.right_sample_block))}
```

Comment:
Write to right_file prepared blocks, so they are ready to insert into hash table and no need to restructure them